bin/ethdl/src/main.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2015-2020 Parity Technologies (UK) Ltd.
+// This file is part of OpenEthereum.
+
+// OpenEthereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// OpenEthereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with OpenEthereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use std::path::Path;
+use std::error::Error;
+use structopt::StructOpt;
+use std::time::SystemTime;
+use tokio_stream::{self as tstream};
+use futures::stream::StreamExt;
+use indicatif::{ProgressBar, ProgressStyle};
+
+#[derive(Debug, StructOpt)]
+#[structopt(name="ethdl")]
+struct DownloadOptions {
+  
+  #[structopt(short, long)]
+  start_block: Option<u64>,
+  
+  #[structopt(short, long)]
+  end_block: Option<u64>,
+
+  #[structopt()]
+  apikey: String,
+
+  #[structopt(short= "n", long="network", default_value="kovan")]
+  network: String,
+
+  #[structopt(short= "o", long="output", default_value=".")]
+  output_dir: String
+}
+
+fn request_url(options: &DownloadOptions) -> String {
+  format!("https://{}.infura.io/v3/{}", options.network, options.apikey)
+}
+
+async fn get_blockchain_height(options: &DownloadOptions) 
+  -> Result<u64, Box<dyn std::error::Error>> {
+  let res_json: serde_json::Value = reqwest::Client::new()
+    .post(&request_url(&options))
+    .json(&serde_json::json!({
+      "jsonrpc": "2.0",
+      "id": SystemTime::now().elapsed()?.subsec_nanos(),
+      "method": "eth_blockNumber",
+      "params": []
+    })).send().await?.json().await?;
+  let valuehex = res_json.get("result").unwrap().as_str().unwrap();
+  Ok(u64::from_str_radix(&valuehex[2..], 16)?)
+}
+
+async fn get_block_by_number(options: &DownloadOptions, number: u64) 
+  -> Result<bool, reqwest::Error> {
+  let filename = format!("{}/{}.json", options.output_dir, number);
+  match Path::new(&filename).exists() {
+    true => Ok(false),
+    false => {
+      serde_json::to_writer_pretty(
+        &std::fs::File::create(&filename).unwrap(), 
+        &reqwest::Client::new()
+          .post(&request_url(&options))
+          .json(&serde_json::json!({
+            "jsonrpc": "2.0",
+            "id": SystemTime::now().elapsed().unwrap().subsec_nanos(),
```

Comment:
This seems to be an awkward way to generate a random id, `.elapsed().unwrap()` might even crash. Why not use a library PRNG? (It's not that important for a small tool like this though)

Reply:
AFAIK, this id is only used by the client, the server just copies this value in the response. So, in this case this is not especially relevant.

Reply:
@jomuel That's right, I was just looking for a quick one-liner to use here without having to have a global static random number generator object. For "real" code this is obviously not a good idea what I'm doing here.

@adria0 I wasn't aware of the implementation details and just used a random number in case the server tries to correlate responses with requests using this field.

In any case, with Dragan's dump that I'm downloading right now, this code will not be used very much anymore.